fontforge/featurefile.c

```diff
@@ -5489,24 +5504,53 @@ return;
     }
     fea_end_statement(tok);
 
-    /* Now a multiple substitution may have a single destination. In which case*/
-    /*  it will look just like a single substitution. So if there are both */
-    /*  multiple and single subs in a feature, translate all the singles into */
-    /*  multiples */
-    /* Another approach would be to make two lookups, but this is easier... */
-    has_single = has_multiple = false;
+    /* Any of a multiple, alternate, or ligature substitution may have */
+    /* a single destination. In either case it will look just like a single */
+    /* substitution. So if there are both multiple/alternate/ligature and */
+    /* single subs in a lookup translate all the singles into the */
+    /* corresponding type */
+    has_single = has_multiple = has_ligature = has_alternate = false;
     for ( item=tok->sofar ; item!=NULL && item->type!=ft_feat_start; item=item->next ) {
 	enum otlookup_type cur = fea_LookupTypeFromItem(item);
 	if ( cur==gsub_multiple )
 	    has_multiple = true;
+	else if ( cur==gsub_alternate )
+	    has_alternate = true;
+	else if ( cur==gsub_ligature )
+	    has_ligature = true;
 	else if ( cur==gsub_single )
 	    has_single = true;
     }
-    if ( has_multiple && has_single ) {
-	for ( item=tok->sofar ; item!=NULL && item->type!=ft_feat_start; item=item->next ) {
-	    enum otlookup_type cur = fea_LookupTypeFromItem(item);
-	    if ( cur==gsub_single )
-		item->u2.pst->type = pst_multiple;
+    if ( has_single ) {
+	enum possub_type psttype = pst_pair;
+	if ( has_multiple && !( has_alternate || has_ligature ) )
+	    psttype = pst_multiple;
+	else if ( has_alternate && !( has_multiple || has_ligature ) )
+	    psttype = pst_alternate;
+	else if ( has_ligature && !( has_multiple || has_alternate ) )
+	    psttype = pst_ligature;
+	if ( psttype!=pst_pair ) {
+	    for ( item=tok->sofar ; item!=NULL && item->type!=ft_feat_start; item=item->next ) {
+		enum otlookup_type cur = fea_LookupTypeFromItem(item);
+		if ( cur==gsub_single )
+		    item->u2.pst->type = psttype;
+	    }
+	}
+    }
+
+    /* Make sure all entries in this lookup of the same lookup type */
```

Comment:
Are we sure this applies to feature definitions too, and not just lookup definitions? Because this check is causing a feature file which previously parsed OK to now fail ([here](https://github.com/BlackFoundry/InriaFonts/blob/master/masters/INRIA-SANS/InriaSans-Regular.ufo/features.fea#L140))

Reply:
The answer seems to be "no, we're not sure". And looking into this the news is better than I was expecting at first, but it's still not clear what to do. 

Convert the font with `ufo2ft`, output as OpenType, and open in FontForge and you wind up with four `frac` subtables. Interestingly two of those are single substitution tables -- the one for the slashes is separate from the one for numerator substitution. Then there's a ligature table and a contextual table. So the idea seems to be to just dynamically sort out the mixture of subtables needed when reading a feature, rather than restricting it to a given type. 

With FontForge, remove all this code, including the old multiple loop, and open the UFO in FontForge and you get three `frac` subtables. This is the good news -- I'm surprised there's any "sorting" going on at all. The result looks similar to the OTF except that there's only one table for single substitutions.

At first glance, at least, both seem to work in the Metrics window. 

But here's the problem: FontForge seems to just be sorting the entries by type and then putting them in *some* order -- maybe the order in which a given type is first encountered. `ufo2ft`, and therefore probably fontTools, seems to be creating a new subtable each time there is a switch in type. That's likely to cause different results in some cases, although I'm not positive about that. So CW was right to leave out the check here, and I was wrong to add it back in, but it's not yet clear to me what changes are needed. 

@khaledhosny can you give any advice on this?

Reply:
IndraSerif-Regular's 8 is ugly, so it may not be necessary to fix this.

Reply:
[Aha](https://adobe-type-tools.github.io/afdko/OpenTypeFeatureFileSpecification.html#7b-ordering-of-lookups-and-subtables):
> A lookup in the OpenType font will be created from each named lookup block (§4.e) or each run of rules with the same feature, script, language, lookupflag and lookup type attribute.

So neither the old or my replacement code was right -- it should all be removed. But there's a bit more work to do in either the parsing or application stage, because this isn't quite what FontForge is doing.